Content for Introduction to Testing and Testing Python code chapter (#77)
Simulation-Software-Engineering/Lecture-Material: 05_testing_and_ci/intro_slides.md

---
type: slide
slideOptions:
  transition: slide
  width: 1400
  height: 900
  margin: 0.1
---

<style>
  .reveal strong {
    font-weight: bold;
    color: orange;
  }
  .reveal p {
    text-align: left;
  }
  .reveal section h1 {
    color: orange;
  }
  .reveal section h2 {
    color: orange;
  }
</style>

> # Introduction to Testing

Author: This slide is the title of this part of the presentation (single `#`) hence the capital case. The rest are titles of slides (`##`).

> # Introduction to Testing

Reviewer: I just wanted to mention it since I was not sure if this difference was on purpose. In the end the title slide is also a slide (to me).

---

## Learning goals of the chapter

- Motivation and importance of writing tests for a software
- Understanding the different types of tests and their relevance in testing scientific software
- Becoming familiar with Python testing frameworks `pytest` and `unittest`

---

## Lets start with a survey

- survey link

> - slido ID

Author: I have a sketch of it in my mind, will have the survey up today or tomorrow and will trigger you to have a look then.

---

## What is testing?

- Smelling old milk before using it!
- A way to determine if a software is producing reliable results
- Manual testing vs. Automated testing

---

## Why should you test your software?

> - Improve software reliability and reproducibility

Author: Good point, I'll mention some examples here

> - Almost all software is better off being tested regularly

Author: Actually no.. I'll be more assertive here :+1:

- Ensure that a distributed version of a software actually works

---

## Nomenclature in software testing

- **Fixture**: preparatory set for testing
- **Actual result**: what the code produces when given the fixture
- **Expected result**: what the actual result is compared to
- **Test coverage**: How much of the code does a test touch

---

## Ways to test software

- Assertions
- Unit testing
- Integration testing
- Regression testing

---

> ## Assertions

Author: The general principle I follow is that if a bullet point has a phrase or just keywords then I do not end it with a full stop at the end. I only use the full stop punctuation. I will make this consistent.

---
type: slide
slideOptions:
  transition: slide
  width: 1400
  height: 900
  margin: 0.1
---

<style>
  .reveal strong {
    font-weight: bold;
    color: orange;
  }
  .reveal p {
    text-align: left;
  }
  .reveal section h1 {
    color: orange;
  }
  .reveal section h2 {
    color: orange;
  }
</style>

# Introduction to Testing

---

> ## Learning goals of the chapter

Reviewer 2: @ajaust and me always (most of the time) used title case for slides headings ... we need a style guide eventually :D

- Understanding why and how to write tests
- Understanding the different types of tests and their relevance in testing scientific software
- Becoming familiar with Python testing frameworks `pytest` and `unittest`

---

## Lets start with a survey

- survey link
- slido ID

---

## What is testing?

- Smelling old milk before using it!
- A way to determine if a software is not producing reliable results and if so, what is the reason.
- Manual testing vs. Automated testing.

---

## Why should you test your software?

- Improve software reliability and reproducibility.
- Make sure that unrelated changes (bugfixes, new features) do not affect other parts of software.
- Generally all software is better off being tested regularly. Possible exceptions are very small codes with single users.
- Ensure that a distributed version of a software actually works.

---

## Nomenclature in software testing

- **Fixture**: preparatory set for testing
- **Actual result**: what the code produces when given the fixture
- **Expected result**: what the actual result is compared to

> - **Test coverage**: how much of the code does a test touch

Reviewer 2: ```suggestion
- **Test coverage**: how much of the code do tests touch
```
?

---

## Some ways to test software

- Assertions
- Unit testing
- Integration testing
- Regression testing

---

## Assertions

- Principle of *defensive programming*.

> - Python does nothing when an assertion is true; returns error when false.

Reviewer 2: This is true for any language, no? But I would say "throw error" instead of "return error".

- Types of assertion statements:
    - Precondition
    - Postcondition
    - Invariant
- A basic but powerful tool to test a software on-the-go.
- Assertion statement syntax:

```python
assert condition, "message"
```

---

## Unit testing

- Catching errors with assertions is good but preventing them is better!

> - A *unit* is a single function in one situation.

Reviewer 2: What do you mean by *one* situation here?

Author: For example a function has numeric input parameters. One situation would be to have integer quantities as inputs and another situation would be to have real parameters as inputs. All combinations of input parameters form a set of situations in which the test is run.

- User creates the expected result manually.

> - A fixture is used to generate an actual result.

Reviewer 2: ```suggestion
- A fixture is used to generate input data.
```
?

> - Assertion statement is used to compare expected result and actual result.

Reviewer 2: I think this is only true for the Python frameworks you had a look at. In general, one would not use an assertion here, but a proper *test* function.

---

## Integration testing

- Test whether several units work in conjunction.
- *Integrate* units and test them together in an *integration* test.
- Often more complicated than a unit test and has more test coverage.
- A fixture is used to generate an actual result.

> - Assertion statement is used to compare expected result and actual result.

Reviewer 2: See above

---

## Regression testing

- Generating an expected result is not possible in some situations.
- Compare the current actual result with a previous actual result.
- No guarantee that the current actual result is correct.
- Risk of a bug being carried over indefinitely.
- Main purpose is to identify changes in the current state of the code.

---

> ## Some important points

Reviewer 2: ```suggestion
## Further remarks
```

- Aim for high test coverage.
- Comparison of floating point variables needs to be done to a certain tolerance.
- Comparing visualizations is often very sensitive, it is better to compare the underlying data instead.
- Test-driven development (TDD) vs. Checking-driven development (CDD).
- Validation vs. Verification from the perspective of testing scientific software.

> - Always make sure that a test breaks when it is supposed to.

Reviewer 2: This last point sounds obvious. What do you want to say here?

Author: I want to convey that when a test is written and if it passes right away then there is no guarantee that the test is correct. Only when a user purposely breaks a test can they be sure that the test works as expected. I will stress on this more explicitly during the lecture.

> 

Reviewer 2: Another point one could add is that whenever you fix a bug in a software, write a test that would have caught the bug. To safeguard against running into the same bug again.

Author: What you say is what I meant by the last point, but I think what I am saying is indirect. I will be more direct here.
